### ISSUE TITLE
Not being able to save T5-11B checkpoint using deepspeed

### ISSUE THREAD

**Issue by tuhinjubcse** [microsoft/DeepSpeed#1686]
**Describe the bug**
Not being able to save T5-11B checkpoint using deepspeed

**To Reproduce**

```
export BS=12;
PYTHONPATH=../../../src
USE_TF=0

deepspeed --num_gpus=4 ./run_translation.py \
        --model_name_or_path  /local/nlp/temp/poetryT511B0/checkpoint-801 \
        --output_dir /local/nlp/temp/poetryT511B1 \
        --evaluation_strategy=steps \
        --save_strategy=epoch \
        --eval_steps 200 \
        --save_steps 200 \
        --do_train \
        --do_eval \
        --train_file /home/tuhin.chakr/gpt3/poetrynew/train.json \
        --validation_file /home/tuhin.chakr/gpt3/poetrynew/val.json \
        --learning_rate 1e-3 \
        --gradient_accumulation_steps 21 \
        --overwrite_output_dir \
        --max_source_length 64 \
        --max_target_length 64 \
        --num_train_epochs 1 \
        --per_device_train_batch_size $BS \
        --per_device_eval_batch_size $BS \
        --source_lang en_XX \
        --target_lang en_XX \
        --deepspeed /home/tuhin.chakr/gpt3/transformers/tests/deepspeed/ds_config_zero3_1.json
```

**Expected behavior**
save checkpoint after training

**ds_report output**
```
[INFO|trainer.py:2250] 2022-01-10 09:42:18,771 >>   Num examples = 65394
[INFO|trainer.py:2253] 2022-01-10 09:42:18,771 >>   Batch size = 12
{'eval_loss': 1.306259274482727, 'eval_runtime': 1585.6465, 'eval_samples_per_second': 41.241, 'eval_steps_per_second': 0.86, 'epoch': 1.0}                                                                                                                                                                                                                          
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 801/801 [20:22:08<00:00, 559.84s/it][INFO|trainer.py:2003] 2022-01-10 10:10:10,357 >> Saving model checkpoint to /local/nlp/temp/poetryT511B1/checkpoint-801                                                                                                                                                                                                                                             
[INFO|configuration_utils.py:423] 2022-01-10 10:10:10,358 >> Configuration saved in /local/nlp/temp/poetryT511B1/checkpoint-801/config.json
[INFO|modeling_utils.py:1070] 2022-01-10 10:10:10,516 >> Model weights saved in /local/nlp/temp/poetryT511B1/checkpoint-801/pytorch_model.bin
[INFO|tokenization_utils_base.py:2043] 2022-01-10 10:10:10,517 >> tokenizer config file saved in /local/nlp/temp/poetryT511B1/checkpoint-801/tokenizer_config.json
[INFO|tokenization_utils_base.py:2049] 2022-01-10 10:10:10,517 >> Special tokens file saved in /local/nlp/temp/poetryT511B1/checkpoint-801/special_tokens_map.json
[INFO|tokenization_t5_fast.py:159] 2022-01-10 10:10:10,566 >> Copy vocab file to /local/nlp/temp/poetryT511B1/checkpoint-801/spiece.model
Traceback (most recent call last):
  File "./run_translation.py", line 626, in <module>
Traceback (most recent call last):
Traceback (most recent call last):
  File "./run_translation.py", line 626, in <module>
  File "./run_translation.py", line 626, in <module>
Traceback (most recent call last):
  File "./run_translation.py", line 626, in <module>
    main()
  File "./run_translation.py", line 543, in main
    main()
  File "./run_translation.py", line 543, in main
    main()
  File "./run_translation.py", line 543, in main
    main()
  File "./run_translation.py", line 543, in main
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1399, in train
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1399, in train
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1399, in train
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1399, in train
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1503, in _maybe_log_save_evaluate
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1503, in _maybe_log_save_evaluate
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1503, in _maybe_log_save_evaluate
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1503, in _maybe_log_save_evaluate
    self._save_checkpoint(model, trial, metrics=metrics)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1565, in _save_checkpoint
    self._save_checkpoint(model, trial, metrics=metrics)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1565, in _save_checkpoint
    self._save_checkpoint(model, trial, metrics=metrics)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1565, in _save_checkpoint
    self._save_checkpoint(model, trial, metrics=metrics)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1565, in _save_checkpoint
    self.save_model(output_dir)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1966, in save_model
    self.save_model(output_dir)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1966, in save_model
    self.save_model(output_dir)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1966, in save_model
    self.deepspeed.save_fp16_model(output_dir, WEIGHTS_NAME)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 3024, in save_fp16_model
    self.save_model(output_dir)
  File "/home/tuhin.chakr/yes/envs/fairseq/lib/python3.8/site-packages/transformers/trainer.py", line 1966, in save_model
        self.deepspeed.save_fp16_model(output_dir, WEIGHTS_NAME)self.deepspeed.save_fp16_model(output_dir, WEIGHTS_NAME)

  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 3024, in save_fp16_model
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 3024, in save_fp16_model
    self.deepspeed.save_fp16_model(output_dir, WEIGHTS_NAME)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 3024, in save_fp16_model
    state_dict = self._zero3_consolidated_fp16_state_dict()
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2999, in _zero3_consolidated_fp16_state_dict
    state_dict = self._zero3_consolidated_fp16_state_dict()
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2999, in _zero3_consolidated_fp16_state_dict
    state_dict = self._zero3_consolidated_fp16_state_dict()
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2999, in _zero3_consolidated_fp16_state_dict
    state_dict = self._zero3_consolidated_fp16_state_dict()
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2999, in _zero3_consolidated_fp16_state_dict
    get_layer_state_dict(self.module, prefix="")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(self.module, prefix="")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(self.module, prefix="")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(self.module, prefix="")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  [Previous line repeated 4 more times]
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2991, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  [Previous line repeated 4 more times]
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2991, in get_layer_state_dict
    state_dict[prefix + name] = buf.detach().cpu()
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 1327, in __exit__
    get_layer_state_dict(child, prefix + name + ".")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    state_dict[prefix + name] = buf.detach().cpu()
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 1327, in __exit__
    get_layer_state_dict(child, prefix + name + ".")
  [Previous line repeated 4 more times]
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2991, in get_layer_state_dict
    state_dict[prefix + name] = buf.detach().cpu()
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 1327, in __exit__
    get_layer_state_dict(child, prefix + name + ".")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2996, in get_layer_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  [Previous line repeated 4 more times]
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/engine.py", line 2991, in get_layer_state_dict
    state_dict[prefix + name] = buf.detach().cpu()
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 1327, in __exit__
        self.params[0].partition(param_list=self.params, has_been_updated=True)self.params[0].partition(param_list=self.params, has_been_updated=True)

  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 604, in partition
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 604, in partition
        self.params[0].partition(param_list=self.params, has_been_updated=True)self.params[0].partition(param_list=self.params, has_been_updated=True)

  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 604, in partition
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 604, in partition
    self._partition(param_list, has_been_updated=has_been_updated)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 716, in _partition
    self._partition(param_list, has_been_updated=has_been_updated)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 716, in _partition
    self._partition(param_list, has_been_updated=has_been_updated)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 716, in _partition
    self._partition(param_list, has_been_updated=has_been_updated)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 716, in _partition
    self._partition_param(param, has_been_updated=has_been_updated)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 725, in _partition_param
    self._partition_param(param, has_been_updated=has_been_updated)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 725, in _partition_param
    self._partition_param(param, has_been_updated=has_been_updated)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 725, in _partition_param
    self._partition_param(param, has_been_updated=has_been_updated)
  File "/home/tuhin.chakr/DeepSpeed/deepspeed/runtime/zero/partition_parameters.py", line 725, in _partition_param
    assert param.ds_status is not ZeroParamStatus.INFLIGHT, f" {param} Cannot partition a param in flight"
AssertionError:  Parameter containing:
tensor([[-0.0184,  0.0311,  0.0164,  ...,  0.0964,  0.0053,  0.0294],
        [ 0.0060, -0.0118,  0.0124,  ..., -0.0006,  0.0004,  0.0281],
        [-0.0068,  0.0219, -0.0637,  ...,  0.0357,  0.0150,  0.0212],
        ...,
        [ 0.0526, -0.0020,  0.0183,  ...,  0.0039,  0.0156,  0.0289],
        [ 0.0212, -0.0099, -0.0158,  ...,  0.0561,  0.0485,  0.0107],
        [ 0.0658,  0.0129,  0.1380,  ..., -0.0192, -0.0014,  0.0330]],
       device='cuda:2', requires_grad=True) Cannot partition a param in flight
    assert param.ds_status is not ZeroParamStatus.INFLIGHT, f" {param} Cannot partition a param in flight"
AssertionError:  Parameter containing:
tensor([[-0.0184,  0.0311,  0.0164,  ...,  0.0964,  0.0053,  0.0294],
        [ 0.0060, -0.0118,  0.0124,  ..., -0.0006,  0.0004,  0.0281],
        [-0.0068,  0.0219, -0.0637,  ...,  0.0357,  0.0150,  0.0212],
        ...,
        [ 0.0526, -0.0020,  0.0183,  ...,  0.0039,  0.0156,  0.0289],
        [ 0.0212, -0.0099, -0.0158,  ...,  0.0561,  0.0485,  0.0107],
        [ 0.0658,  0.0129,  0.1380,  ..., -0.0192, -0.0014,  0.0330]],
       device='cuda:1', requires_grad=True) Cannot partition a param in flight
    assert param.ds_status is not ZeroParamStatus.INFLIGHT, f" {param} Cannot partition a param in flight"
AssertionError:  Parameter containing:
tensor([[-0.0184,  0.0311,  0.0164,  ...,  0.0964,  0.0053,  0.0294],
        [ 0.0060, -0.0118,  0.0124,  ..., -0.0006,  0.0004,  0.0281],
        [-0.0068,  0.0219, -0.0637,  ...,  0.0357,  0.0150,  0.0212],
        ...,
        [ 0.0526, -0.0020,  0.0183,  ...,  0.0039,  0.0156,  0.0289],
        [ 0.0212, -0.0099, -0.0158,  ...,  0.0561,  0.0485,  0.0107],
        [ 0.0658,  0.0129,  0.1380,  ..., -0.0192, -0.0014,  0.0330]],
       device='cuda:3', requires_grad=True) Cannot partition a param in flight    
assert param.ds_status is not ZeroParamStatus.INFLIGHT, f" {param} Cannot partition a param in flight"
AssertionError:  Parameter containing:
tensor([[-0.0184,  0.0311,  0.0164,  ...,  0.0964,  0.0053,  0.0294],
        [ 0.0060, -0.0118,  0.0124,  ..., -0.0006,  0.0004,  0.0281],
        [-0.0068,  0.0219, -0.0637,  ...,  0.0357,  0.0150,  0.0212],
        ...,
        [ 0.0526, -0.0020,  0.0183,  ...,  0.0039,  0.0156,  0.0289],
        [ 0.0212, -0.0099, -0.0158,  ...,  0.0561,  0.0485,  0.0107],
        [ 0.0658,  0.0129,  0.1380,  ..., -0.0192, -0.0014,  0.0330]],
       device='cuda:0', requires_grad=True) Cannot partition a param in flight
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████

```

**Comment by stas00** [comment]
Thank you for the report, @tuhinjubcse 

That's interesting. It means that some gpus are still syncing their data and which means we need some sort of syncing barrier before we can safely save the model weights.

I don't see anything of sorts in:

https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L2639

May be this:

https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L2656

but I don't see any references to communication sync. 

I found a zero-3 specific way:

https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage3.py#L450-L451

which is probably what needs to be called. But why is it not called from `save_checkpoint`?

@tjruwase, any idea what we should add at the top of:

https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L3017

to ensure params in flights have landed?

Thanks.

**Comment by tuhinjubcse** [comment]
![image](https://user-images.githubusercontent.com/3104771/148947942-93f9120d-2d54-46a1-b98c-2ef942428a54.png)
It behaves randomly, it could save when i tried to save the checkpoint after training 50% of the epoch

**Comment by tjruwase** [comment]
@tuhinjubcse, thanks for reporting this problem. Have you found a way to reliably repro? Can you share how to do so, including the relevant config/json files? Thanks!

**Comment by tuhinjubcse** [comment]
```
Name: deepspeed
Version: 0.5.8+09260b6
Summary: DeepSpeed library
Home-page: http://deepspeed.ai
Author: DeepSpeed Team
Author-email: deepspeed@microsoft.com
```


```
{
  
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },

    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto"
        }
    },

    "zero_optimization": {
        "stage": 3,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
        "offload_param": {
            "device": "cpu",
            "pin_memory": true
        },
        "overlap_comm": true,
        "contiguous_gradients": true,
        "sub_group_size": 1e9,
        "reduce_bucket_size": "auto",
        "stage3_prefetch_bucket_size": "auto",
        "stage3_param_persistence_threshold": "auto",
        "stage3_max_live_parameters": 1e9,
        "stage3_max_reuse_distance": 1e9,
        "stage3_gather_fp16_weights_on_model_save": true
    },

    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "steps_per_print": 2000,
    "train_batch_size": 1008,
    "train_micro_batch_size_per_gpu": 12,
    "wall_clock_breakdown": false
}
```

**Comment by tjruwase** [comment]
@tuhinjubcse, thanks for sharing the ds config. How about the following? Are they the same as default settings? 
```bash
        --model_name_or_path  /local/nlp/temp/poetryT511B0/checkpoint-801 \
      --train_file /home/tuhin.chakr/gpt3/poetrynew/train.json \
        --validation_file /home/tuhin.chakr/gpt3/poetrynew/val.json \
```

**Comment by tuhinjubcse** [comment]
Yes, the train and validation files are proper. I have been using the same for prior runs or w/o deep speed

`/local/nlp/temp/poetryT511B0/checkpoint-801 `is a checkpoint trained using deep speed itself

**Comment by tjruwase** [comment]
@tuhinjubcse, sorry that I was not clear on my ask. I am trying to repro this error on my side and those files settings that I highlighted from the command line appear not to be publicly available so that I run the command. It appears that you are running the HF translation example and so I am looking [here](https://github.com/huggingface/transformers/tree/master/examples/pytorch/translation). It appears that both [--train_file](https://github.com/huggingface/transformers/blob/28e091430eea9e0d40839e56fd0d57aec262f5f9/examples/pytorch/translation/run_translation.py#L115) and [--validation_file](https://github.com/huggingface/transformers/blob/28e091430eea9e0d40839e56fd0d57aec262f5f9/examples/pytorch/translation/run_translation.py#L116) are optional to [run_translation.py](https://github.com/huggingface/transformers/blob/master/examples/pytorch/translation/run_translation.py). So, do I need these two settings in order to repro? If so, can you share some version of these two files for my repro effort. 

Regarding '--model_name_or_path', are you able to repro if you simply set that t5-11b? I am curious if the problem is linked to loading the deepspeed checkpoint? Thanks!

**Comment by tuhinjubcse** [comment]
No it's not a problem to load the checkpoint as you see it can save at 400 steps when its 800 steps for 1 epoch

Regarding '--model_name_or_path', are you able to repro if you simply set that to t5-11b? I am curious if the problem is linked to loading the deep speed checkpoint? Thanks!

I could train 1 epoch using t5-11b , but then I wanted to retrain, what I fail to understand if it could save a chekpoint at step 400 why not at 800. I mean I understand you are speculating but does not sound logical to me. 

I can't share the files publicly can I email you ?




```
Using /home/tuhin.chakr/.cache/torch_extensions/py38_cu113 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.0013186931610107422 seconds
[INFO|trainer.py:1196] 2022-01-10 19:06:28,907 >> ***** Running training *****
[INFO|trainer.py:1197] 2022-01-10 19:06:28,907 >>   Num examples = 808180
[INFO|trainer.py:1198] 2022-01-10 19:06:28,907 >>   Num Epochs = 1
[INFO|trainer.py:1199] 2022-01-10 19:06:28,907 >>   Instantaneous batch size per device = 12
[INFO|trainer.py:1200] 2022-01-10 19:06:28,907 >>   Total train batch size (w. parallel, distributed & accumulation) = 1008
[INFO|trainer.py:1201] 2022-01-10 19:06:28,907 >>   Gradient Accumulation steps = 21
[INFO|trainer.py:1202] 2022-01-10 19:06:28,907 >>   Total optimization steps = 801
Using /home/tuhin.chakr/.cache/torch_extensions/py38_cu113 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.004250049591064453 seconds
 12%|███████████████████████▏                                                                                                                                                                           | 95/801 [2:15:09<16:36:08, 84.66s/it][2022-01-10 21:21:56,850] [INFO] [timer.py:181:stop] 0/2000, SamplesPerSec=11.84671345307211
 24%|██████████████████████████████████████████████                                                                                                                                                    | 190/801 [4:28:51<14:19:24, 84.39s/it][2022-01-10 23:35:57,455] [INFO] [timer.py:181:stop] 0/4000, SamplesPerSec=11.902172409886894
 36%|█████████████████████████████████████████████████████████████████████                                                                                                                             | 285/801 [6:42:37<12:06:29, 84.48s/it][2022-01-11 01:50:02,844] [INFO] [timer.py:181:stop] 0/6000, SamplesPerSec=11.91840064624584
 47%|████████████████████████████████████████████████████████████████████████████████████████████▌                                                                                                      | 380/801 [8:56:14<9:52:49, 84.49s/it][2022-01-11 04:03:58,465] [INFO] [timer.py:181:stop] 0/8000, SamplesPerSec=11.930109890839498
 50%|█████████████████████████████████████████████████████████████████████████████████████████████████▍                                                                                                 | 400/801 [9:24:28<9:28:27, 85.06s/it][INFO|trainer.py:2248] 2022-01-11 04:30:57,897 >> ***** Running Evaluation *****
[INFO|trainer.py:2250] 2022-01-11 04:30:57,897 >>   Num examples = 65394
[INFO|trainer.py:2253] 2022-01-11 04:30:57,897 >>   Batch size = 12
{'eval_loss': 1.4324545860290527, 'eval_runtime': 1635.6359, 'eval_samples_per_second': 39.981, 'eval_steps_per_second': 0.833, 'epoch': 0.5}                                                                                                 
 50%|█████████████████████████████████████████████████████████████████████████████████████████████████▍                                                                                                 | 400/801 [9:51:44<9:28:27, 85.06s/it[INFO|trainer.py:2003] 2022-01-11 04:58:13,535 >> Saving model checkpoint to /local/nlp/temp/poetryT511B1/checkpoint-400                                                                                                                       
[INFO|configuration_utils.py:423] 2022-01-11 04:58:13,536 >> Configuration saved in /local/nlp/temp/poetryT511B1/checkpoint-400/config.json
[INFO|modeling_utils.py:1070] 2022-01-11 04:58:13,568 >> Model weights saved in /local/nlp/temp/poetryT511B1/checkpoint-400/pytorch_model.bin
[INFO|tokenization_utils_base.py:2043] 2022-01-11 04:58:13,568 >> tokenizer config file saved in /local/nlp/temp/poetryT511B1/checkpoint-400/tokenizer_config.json
[INFO|tokenization_utils_base.py:2049] 2022-01-11 04:58:13,568 >> Special tokens file saved in /local/nlp/temp/poetryT511B1/checkpoint-400/special_tokens_map.json
[INFO|tokenization_t5_fast.py:159] 2022-01-11 04:58:13,629 >> Copy vocab file to /local/nlp/temp/poetryT511B1/checkpoint-400/spiece.model
[2022-01-11 04:58:30,520] [INFO] [engine.py:3036:save_fp16_model] Saving model weights to /local/nlp/temp/poetryT511B1/checkpoint-400/pytorch_model.bin
[2022-01-11 04:59:27,965] [INFO] [logging.py:69:log_dist] [Rank 0] Saving model checkpoint: /local/nlp/temp/poetryT511B1/checkpoint-400/global_step400/zero_pp_rank_0_mp_rank_00_model_states.pt
[2022-01-11 05:00:31,949] [INFO] [engine.py:2934:_save_zero_checkpoint] zero checkpoint saved /local/nlp/temp/poetryT511B1/checkpoint-400/global_step400/zero_pp_rank_0_mp_rank_00_optim_states.pt
 59%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎                                                                              | 476/801 [11:39:33<7:35:08, 84.02s/it][2022-01-11 06:46:16,339] [INFO] [timer.py:181:stop] 0/10000, SamplesPerSec=11.958888956974398
{'loss': 1.1626, 'learning_rate': 0.001, 'epoch': 0.62}                                                                                                                                                                                       
 71%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎                                                       | 571/801 [13:53:51<5:25:04, 84.80s/it][2022-01-11 09:00:53,066] [INFO] [timer.py:181:stop] 0/12000, SamplesPerSec=11.949745669402043
 83%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▎                                | 666/801 [16:07:35<3:08:07, 83.61s/it][2022-01-11 11:14:55,904] [INFO] [timer.py:181:stop] 0/14000, SamplesPerSec=11.95049789802465
 95%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▏         | 761/801 [18:21:29<56:03, 84.08s/it][2022-01-11 13:29:10,300] [INFO] [timer.py:181:stop] 0/16000, SamplesPerSec=11.948915900920845
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊| 800/801 [19:16:16<01:24, 84.56s/it][INFO|trainer.py:2248] 2022-01-11 14:22:45,823 >> ***** Running Evaluation *****
[INFO|trainer.py:2250] 2022-01-11 14:22:45,823 >>   Num examples = 65394
[INFO|trainer.py:2253] 2022-01-11 14:22:45,823 >>   Batch size = 12
{'eval_loss': 1.3062974214553833, 'eval_runtime': 1613.0927, 'eval_samples_per_second': 40.54, 'eval_steps_per_second': 0.845, 'epoch': 1.0}                                                                                                  
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▊| 800/801 [19:43:09<01:24, 84.56s/it[INFO|trainer.py:2003] 2022-01-11 14:49:38,918 >> Saving model checkpoint to /local/nlp/temp/poetryT511B1/checkpoint-800                                                                                                                       
[INFO|configuration_utils.py:423] 2022-01-11 14:49:38,919 >> Configuration saved in /local/nlp/temp/poetryT511B1/checkpoint-800/config.json
[INFO|modeling_utils.py:1070] 2022-01-11 14:49:38,950 >> Model weights saved in /local/nlp/temp/poetryT511B1/checkpoint-800/pytorch_model.bin
[INFO|tokenization_utils_base.py:2043] 2022-01-11 14:49:38,951 >> tokenizer config file saved in /local/nlp/temp/poetryT511B1/checkpoint-800/tokenizer_config.json
[INFO|tokenization_utils_base.py:2049] 2022-01-11 14:49:38,951 >> Special tokens file saved in /local/nlp/temp/poetryT511B1/checkpoint-800/special_tokens_map.json
[INFO|tokenization_t5_fast.py:159] 2022-01-11 14:49:39,008 >> Copy vocab file to /local/nlp/temp/poetryT511B1/checkpoint-800/spiece.model
[2022-01-11 14:49:55,856] [INFO] [engine.py:3036:save_fp16_model] Saving model weights to /local/nlp/temp/poetryT511B1/checkpoint-800/pytorch_model.bin
[2022-01-11 14:50:50,341] [INFO] [logging.py:69:log_dist] [Rank 0] Saving model checkpoint: /local/nlp/temp/poetryT511B1/checkpoint-800/global_step800/zero_pp_rank_0_mp_rank_00_model_states.pt
[2022-01-11 14:51:48,922] [INFO] [engine.py:2934:_save_zero_checkpoint] zero checkpoint saved /local/nlp/temp/poetryT511B1/checkpoint-800/global_step800/zero_pp_rank_0_mp_rank_00_optim_states.pt
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 801/801 [19:46:49<00:00, 608.88s/it][INFO|trainer.py:1417] 2022-01-11 14:53:18,121 >> 

Training completed. Do not forget to share your model on huggingface.co/models =)


{'train_runtime': 71209.2142, 'train_samples_per_second': 11.349, 'train_steps_per_second': 0.011, 'train_loss': 1.0901248469930165, 'epoch': 1.0}                                                                                            

```

**Comment by tjruwase** [comment]
@tuhinjubcse, did you respond to my last message? It sounds odd, but I got an email notification of a response but can't find the text here.

**Comment by tuhinjubcse** [comment]
It keeps happening , why would be a problem with loading a checkpoint, it keeps training and then fails after finishing an epoch because it cannot save. I can email you the files . I cannot post it publicly. Do you have an email id ?

**Comment by stas00** [comment]
@tuhinjubcse, to diagnose this quickly what happens if you add:

```
import time
sleep.time(10)
```
just before the call to save the checkpoint?

If the problem goes away then it's pretty obvious that some sort syncing barrier is needed.

**Comment by aphedges** [comment]
@stas00, I'm experiencing what appears to be the same bug but with T5-3B, and I can trigger it reliably. When you said "just before the call to save the checkpoint", where exactly did you mean? I am using `transformers==4.15.0`, and I tried it with the following diff:

```diff
diff --git a/src/transformers/trainer.py b/src/transformers/trainer.py
index d02d7fca9..e294686b2 100755
--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -1522,6 +1522,8 @@ class Trainer:
             self._report_to_hp_search(trial, epoch, metrics)
 
         if self.control.should_save:
+            import time
+            time.sleep(10)
             self._save_checkpoint(model, trial, metrics=metrics)
             self.control = self.callback_handler.on_save(self.args, self.state, self.control)

```

Putting the `sleep()` in there did not help. Is there a better place to put it, or does this mean a deeper dive is needed? I can try to get my code and data into a more usable form for it to be posted to help debugging and testing.

In case it helps, here is my [`ds_report` output](https://github.com/microsoft/DeepSpeed/files/7902308/ds_report.txt) and my [`ds_config.json`](https://github.com/microsoft/DeepSpeed/files/7902450/ds_config.json.zip) (zipped because GitHub is annoying).

**Comment by stas00** [comment]
Hi Alex,

Thank you for trying my suggestion!

I'd have put it here:

https://github.com/huggingface/transformers/blob/baf1ebe9f033b0d8cd1ca095ba68f6789ff3d505/src/transformers/trainer.py#L1641

just before `deepspeed.save_checkpoint` is called to ensure it's the right place, but I'm pretty sure where you put it is functionally the same.

Let's try one last time as t5-3b as a pretty hefty model and I don't know what hardware you're running it on - could you please try once a longer sleep, say 30 secs - just to invalidate that somehow it's not an intragpu network issue - which is super unlikely but let's check that.

If sleep doesn't work then it is probably a bug in deepspeed and then yes we definitely need to try to figure out how to reproduce it before being able to resolve it.

**Comment by stas00** [comment]
Actually probably a much better test would be to actually put an actual gpu synchronization instead of sleep.

what happens if you put:

```
import torch.distributed as dist
dist.barrier()
```

just before `deepspeed.save_checkpoint` as above?

**Comment by aphedges** [comment]
That's interesting. My script, running with `transformers==4.15.0`, fails at [trainer.py:1587](https://github.com/huggingface/transformers/blob/v4.15.0/src/transformers/trainer.py#L1587), which is before the line you highlighted at [trainer.py:1591](https://github.com/huggingface/transformers/blob/v4.15.0/src/transformers/trainer.py#L1591). The earlier line includes a call to `self.deepspeed.save_fp16_model()` at [trainer.py:2000](https://github.com/huggingface/transformers/blob/v4.15.0/src/transformers/trainer.py#L2000), which is where the error occurs. Given that your link was to a more recent commit than the release, I've switched to installing the latest `master` (<https://github.com/huggingface/transformers/commit/515ed3ad2a11a6b0cd9800b2ad4d3b313fdaea8c>) as an editable install to make this debugging easier. I put your suggested GPU synchronization code right before the call to `deepspeed.save_fp16_model()`:

```diff
diff --git a/src/transformers/trainer.py b/src/transformers/trainer.py
index 4b97f1778..1d6d5410f 100755
--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -2046,6 +2046,8 @@ class Trainer:
                 # now save the real model if stage3_gather_fp16_weights_on_model_save=True
                 # if false it will not be saved.
                 # This must be called on all ranks
+                import torch.distributed as dist
+                dist.barrier()
                 self.deepspeed.save_fp16_model(output_dir, WEIGHTS_NAME)

         elif self.args.should_save:
```


Even with this change, I still get the crash. To make things easier to debug, here is the log: [preTrainT5.15829383.txt](https://github.com/microsoft/DeepSpeed/files/7910472/preTrainT5.15829383.txt). If that is not sufficient to figure out the problem, I can work on creating a minimal reproduction. It would take a while to simplify the code and data, though, and it's slow to verify the bug (it takes over 20 minutes to crash), so I hope it won't be needed.

**Comment by stas00** [comment]
OK, so first of all it's clear that as it's failing in `save_fp16_model` and not `save_checkpoint` and Alex's experiment failed, then it's missing the synchronization, which is not a `barrier`. So most likely it needs this:

https://github.com/microsoft/DeepSpeed/blob/4912e0ad7efcaf97389ae944259aa0e9f331038a/deepspeed/runtime/engine.py#L2726-L2728

@aphedges, could you please paste this code instead of the barrier code you shared above?

**Comment by aphedges** [comment]
I tried it with the following diff:

```diff
diff --git a/src/transformers/trainer.py b/src/transformers/trainer.py
index 4b97f1778..cbee2ee36 100755
--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -2046,6 +2046,9 @@ class Trainer:
                 # now save the real model if stage3_gather_fp16_weights_on_model_save=True
                 # if false it will not be saved.
                 # This must be called on all ranks
+                if self.deepspeed.zero_optimization_partition_weights():
+                    # Prepare for state_dict() by ensuring all parameters are partitioned
+                    self.deepspeed.optimizer.save_checkpoint_prologue()
                 self.deepspeed.save_fp16_model(output_dir, WEIGHTS_NAME)

         elif self.args.should_save:
```

It unfortunately didn't work. Here is the log file: [preTrainT5.15829392.txt](https://github.com/microsoft/DeepSpeed/files/7914550/preTrainT5.15829392.txt). On a related note, is there a good way to make logging from different processes easier to read? The mingled stack traces are a little annoying.

I saw the changes that were made in <https://github.com/microsoft/DeepSpeed/commit/4912e0ad7efcaf97389ae944259aa0e9f331038a> last night. Some of my coworkers have been waiting for bfloat16 support in ZeRO 3 for a while, so they'll be quite happy to see it. I agree that sticking to the DeepSpeed version I've been using for testing is easier.

**Comment by stas00** [comment]
OK, so let's attack this on multiple fronts.

1. to unblock you, you don't need `save_fp16_model` so you can completely disable it by setting:

```
stage3_gather_fp16_weights_on_model_save=False
```
Done! Then use [zero_to_fp32.py](https://huggingface.co/docs/transformers/master/main_classes/deepspeed#getting-the-model-weights-out) to recover weights.

cc: @tuhinjubcse as well.

Please confirm that this works, that is `save_checkpoint` works, which leaves us with just `save_fp16_model` to figure out.

2. given that the 2 of you have gotten the same issue now, it tells me that most likely something has changed in deespseed recently. So if you go back to a deepspeed version say from 2 month ago, does the problem go away?

3. we need to try to figure out how to reproduce this. Since `save_fp16_model` is tested on the HF side at least. So perhaps this is a unique combination that needs to be tested as well.

4. @tjruwase, any suggestions from your deepspeed understanding of what the code needs to do to force the param in flight to land?

**Comment by stas00** [comment]
> On a related note, is there a good way to make logging from different processes easier to read? The mingled stack traces are a little annoying.

Totally agree. I have made an attempt to create a non-iterleaved print here: https://github.com/pytorch/pytorch/pull/61803
but this somehow needs to be hooked into the python exception mechanism. I'm open to suggestions to have best integrate this.

From what I noticed inter-node prints don't interleave - only same node ones do. 

Best to discuss this in a separate issue though if you'd like to open one.

**Comment by aphedges** [comment]
Sorry for taking so long to get back to this, but I've needed to focus on other tasks for the past week. I did some work yesterday and today that should hopefully shed some light on the issue.

My answers to 1) and 2) are related. I discovered this bug when I tried to run a script that no one had run for a month or two. I don't actually need an output model except to verify the script still works after committing a couple of uncommitted configuration changes to the repository. It might not be exactly what was originally used, but changing the batch or input size probably shouldn't be causing a bug like this. I unfortunately don't know which dependencies worked two months ago (it was run by coworkers, who have also since updated their environments), so I have just guessed. I've downgraded to the following versions:

- `deepspeed==0.5.3+30537e7` (released 2021-09-18)
- `torch==1.9.1+cu111` (released 2021-09-22)
- `transformers==4.10.3` (released 2021-09-22)

Unfortunately, it still fails with the same error using these 4-month-old versions: [preTrainT5.15851250.txt](https://github.com/microsoft/DeepSpeed/files/7962721/preTrainT5.15851250.txt). It's possible that I have to go even further back to find versions that work, but I don't know how far that needs to be. I was hoping to run `git bisect` once I found a setup that works, but I haven't been able to find one yet.

I'll see if I can reproduce this problem on the latest recent versions with simpler code and data. I managed to cut the time to failure from ~25 to ~5 minutes by using only 2% of the training data, and it hopefully shouldn't be too difficult to simplify the code enough to upload here.

**Comment by stas00** [comment]
Thank you for investigating this, Alex.

I didn't quite understand the comment about proposal (1) - does it still fail if you disable `save_fp16_model` altogether as suggested and if so what's the trace. We need to isolate whether this is a `save_fp16_model` bug or a more general bug in deepspeed. i.e. that there is some scenario when weights don't sync up.

And yes I think the next stage is to get the deepspeed devs to reproduce the issue. In theory the problem should appear even after one or a few forward paths. You can also create synthetic data by removing the DataLoader and replacing it with just a random data generator - or even feed it the same input again and again, which would be even simpler.

**Comment by aphedges** [comment]
Sorry for not being clear. I did not actually try proposal (1) because I don't actually need to be unblocked to have a working model. I just wanted training to work again. When I read it, I missed the part where you wanted me to confirm `save_checkpoint` works if I disable `save_fp16_model`.

Since my previous comment, I did some more investigation trying to create my minimal reproduction. I installed the latest version of `deepspeed` (technically <https://github.com/microsoft/DeepSpeed/commit/9351266f7888ad3059b0fc2908d9bab721cb84b1>), and I got a different error that that triggered under the same conditions as the error we've been investigating. The error is triggered at <https://github.com/microsoft/DeepSpeed/blob/9351266f7888ad3059b0fc2908d9bab721cb84b1/deepspeed/runtime/zero/partition_parameters.py#L255>

and the error I get is `AssertionError: {'id': 0, 'status': 'AVAILABLE', 'numel': 24674304, 'ds_numel': 24674304, 'shape': (32128, 768), 'ds_shape': (32128, 768), 'requires_grad': True, 'grad_shape': None, 'persist': False, 'active_sub_modules': {213}}`.

During the minimal reproduction work, I was able to figure out that I was able to get the model to train without errors if I removed `--gradient_accumulation_steps 4` from the parameters I was using. It had been added right before the breakage, so it makes sense that removing it fixes the problem.

One of the reasons it's taken so long to get a minimal reproduction is that I was weirdly getting failures under one dataset but not the other. Your comment about creating simplified data inspired me to see what truly is the difference between running with the two datasets, and it turns out a couple of parameters are different. I figured out that I was able to make the non-failing dataset fail if I pass the `--save_strategy epoch` parameter. In addition to helping figure out how to trigger the bug, it's good to know the problem is not data-specific.

I'm putting the finishing touches on the reproduction code. I'll post it tomorrow.

**Comment by stas00** [comment]
Oh! how interesting - is it possible that the checkpoint is attempted to be saved before grad accumulation finished its cycle? i.e. it was caught in the middle of it.

Looking at @tuhinjubcse OP, he uses `--gradient_accumulation_steps 21`, which surely is likely to be caught in the middle if the save interval is not constrained to be grad accum divisible - so perhaps this is the culprit and then to reproduce it should be enough to run 1 iteration with  `--gradient_accumulation_steps 2` followed by save and it should then fail.

And then in the case of @tuhinjubcse - changing the `--save_steps 200` to something like `--save_steps 21*10` should work around this problem. 

Thank you for clarifying that (1) hasn't been tried.

**Comment by stas00** [comment]
Success!

@tjruwase, Alex helped us to find the culprit. The problem happens when `save_interval % grad_accum != 0`

Here is a trivial scenario that reproduces the problem
```
BS=4; PYTHONPATH=src USE_TF=0 /usr/bin/time -v deepspeed --num_gpus 2 examples/pytorch/translation/run_translation.py --model_name_or_path t5-small --output_dir /tmp/zero3 --overwrite_output_dir --max_train_samples 10 --max_eval_samples 10 --max_source_length 128 --max_target_length 128 --val_max_target_length 128 --do_train --do_eval --num_train_epochs 1 --per_device_train_batch_size $BS --per_device_eval_batch_size $BS --learning_rate 3e-3 --warmup_steps 500 --predict_with_generate  --save_steps 1 --eval_steps 1 --group_by_length   --dataset_name wmt16 --dataset_config ro-en --source_lang en --target_lang ro --source_prefix "translate English to Romanian: " --deepspeed tests/deepspeed/ds_config_zero3.json --gradient_accumulation_steps 3
```

I use grad_acum 3 and save interval 1! Boom!

```
[INFO|trainer.py:1263] 2022-01-31 21:14:28,673 >> ***** Running training *****
[INFO|trainer.py:1264] 2022-01-31 21:14:28,673 >>   Num examples = 10
[INFO|trainer.py:1265] 2022-01-31 21:14:28,673 >>   Num Epochs = 1
[INFO|trainer.py:1266] 2022-01-31 21:14:28,673 >>   Instantaneous batch size per device = 4
[INFO|trainer.py:1267] 2022-01-31 21:14:28,673 >>   Total train batch size (w. parallel, distributed & accumulation) = 24
[INFO|trainer.py:1268] 2022-01-31 21:14:28,673 >>   Gradient Accumulation steps = 3
[INFO|trainer.py:1269] 2022-01-31 21:14:28,673 >>   Total optimization steps = 1
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.73s/it][INFO|trainer.py:2114] 2022-01-31 21:14:31,515 >> Saving model checkpoint to /tmp/zero3/checkpoint-1
[INFO|configuration_utils.py:430] 2022-01-31 21:14:31,516 >> Configuration saved in /tmp/zero3/checkpoint-1/config.json
[INFO|modeling_utils.py:1074] 2022-01-31 21:14:31,739 >> Model weights saved in /tmp/zero3/checkpoint-1/pytorch_model.bin
[INFO|tokenization_utils_base.py:2074] 2022-01-31 21:14:31,740 >> tokenizer config file saved in /tmp/zero3/checkpoint-1/tokenizer_config.json
[INFO|tokenization_utils_base.py:2080] 2022-01-31 21:14:31,740 >> Special tokens file saved in /tmp/zero3/checkpoint-1/special_tokens_map.json
[INFO|tokenization_t5_fast.py:162] 2022-01-31 21:14:31,766 >> Copy vocab file to /tmp/zero3/checkpoint-1/spiece.model
Traceback (most recent call last):
  File "examples/pytorch/translation/run_translation.py", line 624, in <module>
    main()
  File "examples/pytorch/translation/run_translation.py", line 541, in main
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/trainer.py", line 1459, in train
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/trainer.py", line 1588, in _maybe_log_save_evaluate
Traceback (most recent call last):
  File "examples/pytorch/translation/run_translation.py", line 624, in <module>
    self._save_checkpoint(model, trial, metrics=metrics)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/trainer.py", line 1656, in _save_checkpoint
    main()
  File "examples/pytorch/translation/run_translation.py", line 541, in main
    train_result = trainer.train(resume_from_checkpoint=checkpoint)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/trainer.py", line 1459, in train
    self.save_model(output_dir, _internal_call=True)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/trainer.py", line 2068, in save_model
    self._maybe_log_save_evaluate(tr_loss, model, trial, epoch, ignore_keys_for_eval)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/trainer.py", line 1588, in _maybe_log_save_evaluate
    if not self.deepspeed.save_fp16_model(output_dir, WEIGHTS_NAME):
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3096, in save_fp16_model
    self._save_checkpoint(model, trial, metrics=metrics)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/trainer.py", line 1656, in _save_checkpoint
    self.save_model(output_dir, _internal_call=True)
  File "/mnt/nvme1/code/huggingface/transformers-master/src/transformers/trainer.py", line 2068, in save_model
    return self.save_16bit_model(save_dir, save_filename)    
if not self.deepspeed.save_fp16_model(output_dir, WEIGHTS_NAME):
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3096, in save_fp16_model
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3122, in save_16bit_model
        state_dict = self._zero3_consolidated_16bit_state_dict()return self.save_16bit_model(save_dir, save_filename)

  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3088, in _zero3_consolidated_16bit_state_dict
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3122, in save_16bit_model
    get_layer_state_dict(self.module, prefix="")
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3085, in get_layer_state_dict
    state_dict = self._zero3_consolidated_16bit_state_dict()
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3088, in _zero3_consolidated_16bit_state_dict
    get_layer_state_dict(child, prefix + name + ".")
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3080, in get_layer_state_dict
    get_layer_state_dict(self.module, prefix="")
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3085, in get_layer_state_dict
    state_dict[prefix + name] = buf.detach().cpu()
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 1653, in __exit__
    get_layer_state_dict(child, prefix + name + ".")
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/engine.py", line 3080, in get_layer_state_dict
    state_dict[prefix + name] = buf.detach().cpu()
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 1653, in __exit__
    self.params[0].partition(param_list=self.params, has_been_updated=True)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 899, in partition
    self._partition(param_list, has_been_updated=has_been_updated)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 1043, in _partition
    self.params[0].partition(param_list=self.params, has_been_updated=True)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 899, in partition
    self._partition_param(param, has_been_updated=has_been_updated)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/utils/nvtx.py", line 11, in wrapped_fn
    self._partition(param_list, has_been_updated=has_been_updated)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 1043, in _partition
    return func(*args, **kwargs)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 1160, in _partition_param
    self._partition_param(param, has_been_updated=has_been_updated)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/utils/nvtx.py", line 11, in wrapped_fn
    return func(*args, **kwargs)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 1160, in _partition_param
    free_param(param)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/utils/nvtx.py", line 11, in wrapped_fn
    return func(*args, **kwargs)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 255, in free_param
    free_param(param)
  File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/utils/nvtx.py", line 11, in wrapped_fn
    assert not param.ds_active_sub_modules, param.ds_summary()
AssertionError    return func(*args, **kwargs)
:   File "/mnt/nvme1/code/github/00optimize/deepspeed/deepspeed/runtime/zero/partition_parameters.py", line 255, in free_param
{'id': 132, 'status': 'AVAILABLE', 'numel': 16435200, 'ds_numel': 16435200, 'shape': (32100, 512), 'ds_shape': (32100, 512), 'requires_grad': True, 'grad_shape': None, 'persist': False, 'active_sub_modules': {111}}
    assert not param.ds_active_sub_modules, param.ds_summary()
AssertionError: {'id': 132, 'status': 'AVAILABLE', 'numel': 16435200, 'ds_numel': 16435200, 'shape': (32100, 512), 'ds_shape': (32100, 512), 'requires_grad': True, 'grad_shape': None, 'persist': False, 'active_sub_modules': {111}}
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:02<00:00,  2.23s/it]
[2022-01-31 21:14:32,954] [INFO] [launch.py:178:sigkill_handler] Killing subprocess 35379
[2022-01-31 21:14:32,955] [INFO] [launch.py:178:sigkill_handler] Killing subprocess 35380
[2022-01-31 21:14:32,955] [ERROR] [launch.py:184:sigkill_handler] ['/home/stas/anaconda3/envs/py38-pt110/bin/python', '-u', 'examples/pytorch/translation/run_translation.py', '--local_rank=1', '--model_name_or_path', 't5-small', '--output_dir', '/tmp/zero3', '--overwrite_output_dir', '--max_train_samples', '10', '--max_eval_samples', '10', '--max_source_length', '128', '--max_target_length', '128', '--val_max_target_length', '128', '--do_train', '--do_eval', '--num_train_epochs', '1', '--per_device_train_batch_size', '4', '--per_device_eval_batch_size', '4', '--learning_rate', '3e-3', '--warmup_steps', '500', '--predict_with_generate', '--save_steps', '1', '--eval_steps', '1', '--group_by_length', '--dataset_name', 'wmt16', '--dataset_config', 'ro-en', '--source_lang', 'en', '--target_lang', 'ro', '--source_prefix', 'translate English to Romanian: ', '--deepspeed', 'tests/deepspeed/ds_config_zero3.json', '--gradient_accumulation_steps', '3'] exits with return code = 1
Command exited with non-zero status 1
```

**Comment by aphedges** [comment]
I've gotten the reproduction down to under 150 lines of code, most of which is not relevant.

Here are the steps to reproduce the bug:

- Download and extract [reproduce_1686.zip](https://github.com/microsoft/DeepSpeed/files/7976041/reproduce_1686.zip)
- Navigate into the directory with the zip file contents
- In a Python environment with `$CUDA_HOME` appropriately set, run `setup.sh`. I used Python 3.9.9 and CUDA 11.1.0, and I'm not sure if it works with other versions
- Run `reproduce.sh`

Here is the console output ([log_2022_02_01_05_10_16.txt](https://github.com/microsoft/DeepSpeed/files/7976054/log_2022_02_01_05_10_16.txt)) and `ds_report` output ([ds_report.txt](https://github.com/microsoft/DeepSpeed/files/7976053/ds_report.txt)).

**Comment by stas00** [comment]
If I set `"stage3_gather_fp16_weights_on_model_save": false` then it no longer fails.

Odd, I thought you have tried the save_pretrained preamble and it didn't work, perhaps I have mixed things up.

Could you please try with this branch? https://github.com/microsoft/DeepSpeed/pull/1741

**Comment by aphedges** [comment]
I'll try it out tomorrow. Hopefully it works!

**Comment by aphedges** [comment]
I'm glad you figured out what the problem was! The fact that saving in the middle of gradient accumulation causes a failure is a really good explanation for why these two parameters together cause a problem. Switching from `--save-stratgy steps` (which defaults to 500 steps) to `--save-strategy epoch` causes the failure because there likely wasn't even 500 steps during training to encounter any issues with.

I have tried out #1741, and I can confirm it works on my minimal example. I'll try it out on my real tasks and verify it works there as well.

I did try copying the `save_pretrained` preamble as part of <https://github.com/microsoft/DeepSpeed/issues/1686#issuecomment-1018590755>. Checking the log file I attached, the line number in the stack trace match up with the line numbers in the diff. I have no clue why that did not work but your PR did.

**Comment by tuhinjubcse** [comment]
@stas00  This is slightly strange but when I do stage3_gather_fp16_weights_on_model_save": false no pytorch_model.bin file is saved. I am not sure is that a solution? I need the model file to be saved

And I was not getting this error but suddenly getting it again . Can you suggest what to do ? This thread is so long and now hard to follow

```
[INFO|trainer.py:2090] 2022-02-02 02:35:41,452 >> Saving model checkpoint to /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3
[INFO|configuration_utils.py:430] 2022-02-02 02:35:41,481 >> Configuration saved in /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3/config.json
[INFO|modeling_utils.py:1074] 2022-02-02 02:35:42,043 >> Model weights saved in /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3/pytorch_model.bin
[INFO|tokenization_utils_base.py:2074] 2022-02-02 02:35:42,191 >> tokenizer config file saved in /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3/tokenizer_config.json
[INFO|tokenization_utils_base.py:2080] 2022-02-02 02:35:42,299 >> Special tokens file saved in /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3/special_tokens_map.json
[INFO|tokenization_t5_fast.py:162] 2022-02-02 02:35:42,434 >> Copy vocab file to /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3/spiece.model
[2022-02-02 02:35:42,512] [INFO] [engine.py:3027:save_fp16_model] Did not save the model /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3/pytorch_model.bin because `stage3_gather_fp16_weights_on_model_save` is False
[2022-02-02 02:35:42,540] [INFO] [logging.py:69:log_dist] [Rank 0] Saving model checkpoint: /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3/global_step3/zero_pp_rank_0_mp_rank_00_model_states.pt

[2022-02-02 02:42:36,142] [INFO] [engine.py:2934:_save_zero_checkpoint] zero checkpoint saved /mnt/nlp_swordfish/tuhin/poetryT511bcheckpoints/checkpoint-3/global_step3/zero_pp_rank_0_mp_rank_00_optim_states.pt

```

**Comment by stas00** [comment]
> I'm glad you figured out what the problem was! The fact that saving in the middle of gradient accumulation causes a failure is a really good explanation for why these two parameters together cause a problem. Switching from `--save-stratgy steps` (which defaults to 500 steps) to `--save-strategy epoch` causes the failure because there likely wasn't even 500 steps during training to encounter any issues with.

Exactly!

> I have tried out #1741, and I can confirm it works on my minimal example. I'll try it out on my real tasks and verify it works there as well.

Excellent. Let us know about the full program.
 
> I did try copying the `save_pretrained` preamble as part of [#1686 (comment)](https://github.com/microsoft/DeepSpeed/issues/1686#issuecomment-1018590755). Checking the log file I attached, the line number in the stack trace match up with the line numbers in the diff. I have no clue why that did not work but your PR did.

perhaps you changed it in the source, but the copy that was run was different. It's all good.

**Comment by tjruwase** [comment]
> I'm glad you figured out what the problem was! The fact that saving in the middle of gradient accumulation causes a failure is a really good explanation for why these two parameters together cause a problem. Switching from `--save-stratgy steps` (which 

This is a really good catch. I had never previously considered the interactions of gradient accumulation and checkpointing.

**Comment by aphedges** [comment]
> Excellent. Let us know about the full program.

The full program works!

> perhaps you changed it in the source, but the copy that was run was different. It's all good.

I was using an editable install of `transformers` and it's pure Python, so it's not like I needed to recompile. I really have no clue. Anyway, if we discovered the fix that way, we probably wouldn't have realized that were was an interaction between gradient accumulation and checkpointing.

> Please let me know if you need help with figuring out how to install deepspeed from a custom branch.

Not extremely relevant, but today I learned that `pip install https://github.com/stas00/DeepSpeed/archive/refs/heads/save_16bit_model-save_checkpoint_prologue.zip` works.

**Comment by stas00** [comment]
There are many ways to do that, typically the common no cloning the repo way is:

```
pip install git+https://github.com/owner/project.git@remote_branch_name
```

So here it'd be:

```
pip install git+https://github.com/stas00/DeepSpeed@save_16bit_model-save_checkpoint_prologue
```

**Comment by aphedges** [comment]
Today I tried out the revised fix that made it into <https://github.com/microsoft/DeepSpeed/commit/baef92e26fef5aa0da63f26d444b91c2a7aa0bd3> on my full script, and it worked properly! Thank you very much for your work on fixing this issue!

---

I am aware of the VCS installation in pip and it's what I usually use, but it seems to be slightly faster to let GitHub zip it first, at least with n=1. I guess it also helps if one doesn't have Git on their system for some reason.

**Comment by stas00** [comment]
I appreciate you re-validating the fix. @aphedges! 

@tjruwase made it much easier to do so as he tirelessly makes the deepspeed codebase easier to adjust so I had changed its final version.

------------

re: VCS

I just have a clone, which is even faster to `git pull` :)

There are many ways to install a module ;)
